rest/interceptors/src/main/java/com/connexta/spring/interceptor/VersionInterceptor.java

```diff
@@ -0,0 +1,290 @@
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.ProxyInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+/**
+ * Spring interceptor capable of validating the version provided by the client with the version
+ * implemented by the server and return an error message accordingly. It will also automatically add
+ * the <code>Content-Version</code> header with the server version for the API called. This only
+ * happens in cases where the controller was called and generated a response and no other
+ * interceptors failed.
+ *
+ * <p>The server version for the API is retrieved from the dependency.proeprties file that must be
+ * included in the generated spring API using Apache's ServiceMix depends maven plugin.
+ */
+@Component
+public class VersionInterceptor implements HandlerInterceptor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(VersionInterceptor.class);
+
+  private static final String ACCEPT_VERSION = "Accept-Version";
+  private static final String CONTENT_VERSION = "Content-Version";
+
+  public static final String DEPENDENCIES_FILE = "META-INF/maven/dependencies.properties";
+  public static final String VERSION = "version";
+
+  private static final Map<Method, String> versions = new ConcurrentHashMap<>();
+
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+      throws IOException {
+    final HandlerMethod handlerMethod = (HandlerMethod) handler;
+    final String serverVersion = VersionInterceptor.getVersion(handlerMethod.getMethod());
+    final String clientVersion = request.getHeader(VersionInterceptor.ACCEPT_VERSION);
+
+    LOGGER.debug(
+        "VersionInterceptor.preHandle(handler: {}, server: {}, client: {}",
+        handlerMethod.getShortLogMessage(),
+        serverVersion,
+        clientVersion);
+    if ((serverVersion != null)
+        && !VersionInterceptor.areCompatible(serverVersion, clientVersion)) {
+      response.setContentType(MediaType.APPLICATION_JSON_UTF8.toString());
+      response.addHeader(VersionInterceptor.CONTENT_VERSION, serverVersion);
+      response.setStatus(HttpServletResponse.SC_NOT_IMPLEMENTED);
+      response
+          .getWriter()
+          .write(
+              new ObjectMapper()
+                  .writeValueAsString(
+                      new ErrorMessage()
+                          .status(HttpServletResponse.SC_NOT_IMPLEMENTED)
+                          .message("Unsupported version: " + clientVersion)
+                          .addDetailsItem("server version: " + serverVersion)
+                          .path(request.getServletPath())));
+      return false;
```

Comment:
❓ Should we also log something?

Reply:
I would prefer to actually change this interceptor to not generate the response directly but instead throw back an exception and leave the response generation out of interceptors. This would allow any server implementation to decide as to how they best wish to translate exceptions (not just this one) into error messages based on their specs. The best place for this to be done is inside exception handlers which sole purpose is to do exactly that. It decouples the logic for validation like we will need for authentication or version validation from how to compose the response.